frontend/src/pages/Match/index.js

```diff
@@ -67,28 +69,23 @@ function MatchPage() {
       setIsMatchFailed(true);
     });
 
-    // Disconnect when unmount
-    return () => socket.disconnect();
-  }, []);
-
-  // Emit match event
-  useEffect(() => {
+     // Emit match event
     if (difficulty && userCtx.username) {
       socket.emit('match', {
         difficulty,
         username: userCtx.username,
       });
     }
-  }, [difficulty, userCtx]);
 
-  // Navigate to room page upon successful match
-  useEffect(() => {
-    if (room && difficulty) {
-      room && navigate('/room', { state: { room, difficulty } });
+    // Unregister listeners when unmount
+    return () => {
+      socket.off('matchSuccess');
+      socket.off('matchFail');
+      socket.off('disconnect');
     }
-  }, [room, difficulty, navigate]);
+  }, [difficulty, navigate, userCtx]);
```

Comment:
Asking for the sake of our deployment cost, does adding `userCtx` causes infinite rendering of this component? If not, ignore this comment

Reply:
Checked, it won't.